File: fsharp/kindergarten-garden/KindergartenGarden.fs
module KindergartenGarden

type Plant =
    | Grass
    | Clover
    | Radishes
    | Violets
    | Unknown

let children =
    [ "Alice"
      "Bob"
      "Charlie"
      "David"
      "Eve"
      "Fred"
      "Ginny"
      "Harriet"
      "Ileana"
      "Joseph"
      "Kincaid"
      "Larry" ]

let parse (code: char) : Plant =
    match code with
    | 'V' -> Violets
    | 'R' -> Radishes
    | 'G' -> Grass
    | 'C' -> Clover
    | _ -> Unknown

let plants (diagram: string) (student: string) : Plant list =
    let index = children |> List.findIndex (fun c -> c = student)

    diagram.Split("\n")
    |> Seq.collect (Seq.skip (index * 2) >> Seq.take 2)
    |> Seq.map parse
    |> Seq.toList

File: cpp/kindergarten-garden/kindergarten_garden.cpp
#include "kindergarten_garden.h"
#include <stdexcept>
#include <unordered_map>
#include <sstream>

namespace kindergarten_garden
{
    Plants get_plant(unsigned char ch)
    {
        switch (ch)
        {
        case 'G':
            return Plants::grass;
            break;
        case 'C':
            return Plants::clover;
            break;
        case 'R':
            return Plants::radishes;
            break;
        case 'V':
            return Plants::violets;
            break;

        default:
            throw std::invalid_argument("Invalid character");
            break;
        }
    }

    std::array<Plants, 4> plants(const std::string &plants, const std::string &student)
    {
        static const std::unordered_map<std::string, int> students{
            {"Alice", 0},
            {"Bob", 1},
            {"Charlie", 2},
            {"David", 3},
            {"Eve", 4},
            {"Fred", 5},
            {"Ginny", 6},
            {"Harriet", 7},
            {"Ileana", 8},
            {"Joseph", 9},
            {"Kincaid", 10},
            {"Larry", 11}};

        int start = students.at(student) * 2;

        std::string row1_;
        std::string row2_;

        std::stringstream ss(plants);
        std::getline(ss, row1_);
        std::getline(ss, row2_);

        return {
            get_plant(row1_[start]),
            get_plant(row1_[start + 1]),
            get_plant(row2_[start]),
            get_plant(row2_[start + 1])};
    }

} // namespace kindergarten_garden
